controllers/storagecluster/odfinfoconfig.go

```diff
@@ -0,0 +1,253 @@
+		odfDeploymentType = OdfDeploymentTypeInternal
+		if isDROptimized, err = getIsDROptimized(r, storageCluster); err != nil {
+			return "", err
+
+		}
+	}
+	var storageSystemName string
+	if storageSystemName, err = getStorageSystemName(storageCluster); err != nil {
+		return "", err
+	}
+	storageClusterCount := len(r.clusters.GetStorageClusters())
+
+	data := OdfInfoData{
+		OdfVersion:          odfVersion,
+		OdfDeploymentType:   odfDeploymentType,
+		StorageClusterCount: storageClusterCount,
+		IsDROptimized:       isDROptimized,
+		StorageSystemName:   storageSystemName,
+		Clients:             []ConnectedClient{},
+		StorageCluster: InfoStorageCluster{
+			NamespacedName:          types.NamespacedName{Name: storageCluster.Name, Namespace: storageCluster.Namespace},
+			StorageProviderEndpoint: storageCluster.Status.StorageProviderEndpoint,
+			CephClusterFSID:         cephFSId,
+		},
+	}
+	yamlData, err := yaml.Marshal(data)
+	if err != nil {
+		return "", err
+	}
+	return string(yamlData), nil
+
+}
+
+func getStorageSystemName(storageCluster *ocsv1.StorageCluster) (string, error) {
+	storageSystemRef := util.Find(storageCluster.OwnerReferences, func(ref *metav1.OwnerReference) bool {
+		return ref.Kind == "StorageSystem"
+	})
+	if storageSystemRef != nil {
+		return storageSystemRef.Name, nil
+	}
+
+	return "", fmt.Errorf(
+		"failed to find parent StorageSystem's name in StorageCluster %q"+
+			" ownerreferences, %v",
+		storageCluster.Name,
+		storageCluster.OwnerReferences)
+
+}
+
+func getIsDROptimized(r *StorageClusterReconciler, storageCluster *ocsv1.StorageCluster) (bool, error) {
+	var cephCluster rookCephv1.CephCluster
+	err := r.Client.Get(
+		r.ctx,
+		types.NamespacedName{
+			Name:      generateNameForCephClusterFromString(storageCluster.Name),
+			Namespace: storageCluster.Namespace},
+		&cephCluster)
+	if err != nil {
+		return false, err
+	}
+	if cephCluster.Status.CephStorage == nil || cephCluster.Status.CephStorage.OSD.StoreType == nil {
+		return false, fmt.Errorf("cephcluster %v status does not have OSD store information, %v", cephCluster.Name, storageCluster.Name)
+	}
+	bluestorerdr, ok := cephCluster.Status.CephStorage.OSD.StoreType["bluestore-rdr"]
+	if !ok {
+		return false, nil
+	}
+	total := getOsdCount(storageCluster, r.serverVersion)
+	if bluestorerdr < total {
+		return false, nil
+	}
+	return true, nil
+}
+
+func getOdfVersion(r *StorageClusterReconciler, storageCluster *ocsv1.StorageCluster) (string, error) {
+	var csvs operatorsv1alpha1.ClusterServiceVersionList
+	err := r.Client.List(r.ctx, &csvs, client.InNamespace(storageCluster.Namespace))
+	if err != nil {
+		return "", err
+	}
+
+	csv := util.Find(csvs.Items, func(csv *operatorsv1alpha1.ClusterServiceVersion) bool {
+		return strings.HasPrefix(csv.Name, OdfOperatorNamePrefix)
+	})
+	if csv != nil {
+		return csv.Spec.Version.String(), nil
+	}
+
+	return "", fmt.Errorf("failed to find csv with prefix %q", OdfOperatorNamePrefix)
```

Comment:
The intention behind this function is to return a value. As such, the return of the value should be the canonical return at the end of the function while the statement returning an error should be the ones that short circuit the execution of the function. 

```suggestion
	if csv == nil {
		return "", fmt.Errorf("failed to find csv with prefix %q", OdfOperatorNamePrefix)
	}

	return csv.Spec.Version.String(), nil
```